File: 0146-lru-cache/0146-lru-cache.py
class LRUCache:

    def __init__(self, capacity: int):
        self.deque=OrderedDict()
        self.cap=capacity

    def get(self, key: int) -> int:
        if key not in self.deque:
            return -1
        self.deque.move_to_end(key)
        return self.deque[key]

    def put(self, key: int, value: int) -> None:
        if key in self.deque:
            self.deque.move_to_end(key)
        self.deque[key]=value
        if len(self.deque)> self.cap:
            self.deque.popitem(last=False)
        
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.cpp
class LRUCache {
public:
    LRUCache(int capacity) {
        cap=capacity;
    }
    
    int get(int key) {
        if (cache.find(key)==cache.end()) return -1;
        usage.splice(usage.begin(),usage,cache[key].second);
        return cache[key].first;
    }
    
    void put(int key, int value) {
        if (cache.find(key)!= cache.end()){
            cache[key].first=value;
            usage.splice(usage.begin(),usage,cache[key].second);
        }
        else{
            if(size(cache)==cap){
                int val=usage.back();
                usage.pop_back();
                cache.erase(val);
            }
            usage.push_front(key);
            cache[key]={value,usage.begin()};
        }
    }
private:
    int cap;
    list<int> usage;
    unordered_map<int,pair<int,list<int>::iterator>> cache;
};


/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */